fvm/tests/default_kernel/dummy.rs

```diff
@@ -0,0 +1,267 @@
+        bs.has(&root).context("failed to load initial state-root")?;
+        bs.has(&manifest_cid)
+            .context("failed to load builtin actor manifest")?;
+
+        // TODO find and document why this needs to be this
+        // TODO V15 requires this and IDK what V16 expects
+        // TODO why is a tuple of (num_actors, manifest) the expected manifest CID?
+        let actors_cid = bs.put_cbor(&(0, manifest_cid), Code::Blake2b256).unwrap();
+
+        // construct state tree from empty root state
+        let state_tree = StateTree::new_from_root(bs, &root)?;
+
+        // generate context from the new generated root and override actors with empty list
+        // TODO should this stay as V15?
+        let ctx = NetworkConfig::new(fvm_shared::version::NetworkVersion::V15)
+            .override_actors(actors_cid)
+            .for_epoch(0, root);
+
+        Ok(Self {
+            ctx,
+            engine: Engine::new_default(EngineConfig {
+                max_wasm_stack: 1024,
+                wasm_prices: fvm::__TESTING_FREE_WASM_PRICE,
+            })?,
+            state_tree,
+            builtin_actors: manifest,
+        })
+    }
+}
+
+impl Machine for DummyMachine {
+    type Blockstore = MemoryBlockstore;
+
+    type Externs = DummyExterns;
+
+    fn engine(&self) -> &Engine {
+        &self.engine
+    }
+
+    fn blockstore(&self) -> &Self::Blockstore {
+        self.state_tree.store()
+    }
+
+    fn context(&self) -> &fvm::machine::MachineContext {
+        &self.ctx
+    }
+
+    fn externs(&self) -> &Self::Externs {
+        &DummyExterns
+    }
+
+    fn builtin_actors(&self) -> &Manifest {
+        &self.builtin_actors
+    }
+
+    fn state_tree(&self) -> &StateTree<Self::Blockstore> {
+        &self.state_tree
+    }
+
+    fn state_tree_mut(&mut self) -> &mut StateTree<Self::Blockstore> {
+        &mut self.state_tree
+    }
+
+    fn create_actor(
+        &mut self,
+        _addr: &Address,
+        _act: ActorState,
+    ) -> kernel::Result<fvm_shared::ActorID> {
+        todo!()
+    }
+
+    fn transfer(
+        &mut self,
+        _from: fvm_shared::ActorID,
+        _to: fvm_shared::ActorID,
+        _value: &fvm_shared::econ::TokenAmount,
+    ) -> kernel::Result<()> {
+        todo!()
+    }
+
+    fn into_store(self) -> Self::Blockstore {
+        todo!()
+    }
+}
+
+#[derive(Deref, DerefMut)]
+// similar to `DefaultCallManager` but with public variables for modification by unit tests
+pub struct InnerDummyCallManager {
+    #[deref]
+    #[deref_mut]
+    pub machine: DummyMachine,
+    pub gas_tracker: GasTracker,
+    pub charge_gas_calls: usize,
+    pub origin: Address,
+    pub nonce: u64,
+}
+/// a wrapper to let us inspect values during testing, all borrows are done with .borrow() or .borrow_mut(), so this should be used in single thereaded tests only
+/// TODO this introduces some rough edges that might need to be cleaned up
+pub struct DummyCallManager(Arc<InnerDummyCallManager>);
```

Comment:
Any reason not to use an `Arc<Mutex>` (or even an `Rc<RefCell>`)? That should avoid having to even think about safety, right?

Reply:
Not really, I was stumbling over `Rc`'s api for too long and decided to go to something I've at least tried once before, I think moving to `Rc<RefCell>` would make most sense here

Reply:
Ah. Ok, so, what if you _don't_ share the entire call manager. I'm pretty sure you don't need to do that.

Instead, store an `Rc<RefCell<TestData>>` inside the `CallManager`, and keep a reference inside the test itself.

1. The call manager will dereference/borrow this to record test operations, check to see if it needs to do something special at some point, etc.
2. The test-case can safely read/write to this shared object.

This should remove a lot of the complexity.

Reply:
well not quite what I am doing currently, but the problem remains where the trait expects a normal reference and `RefCell` will only give out a `Ref`, another option I explored was to keep `Rc` inside of the test and just give the kernel `RefMut<Inner>` though Call manager expects a `'static` lifetime

Reply:
Honestly at this point I think I will avoid doing any fancy instrumentation, and use `into_inner()` after doing some operations, this will get a fair amount of coverage by itself and we can add more complex things later.

Reply:
I'm suggesting:

```rust
struct DummyCallManager {
    field1: Foo,
    field2: Bar,
    the_mutable_stuff: Rc<RefCell<TestData>>,
}
```

You can continue to borrow the outer fields normally, and share/mutate the data in `the_mutable_stuff`.

Reply:
ok, so instead of just keeping each field as shared data, let them be normal but just have another field that stores extra data that can be accessed from the outside. That way the trait can get what it expects and we can do whatever inner mutation in addition to be shared externally

Reply:
Yep. It let's you split out the shared mutable state from the private state.